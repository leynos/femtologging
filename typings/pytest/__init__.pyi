from typing import (
    Any,
    Callable,
    ContextManager,
    Pattern,
    Type,
    TypeVar,
    overload,
    ParamSpec,
)

P = ParamSpec("P")
R = TypeVar("R", bound=BaseException)

@overload
def fixture(func: Callable[P, R]) -> Callable[P, R]: ...
@overload
def fixture(
    *,
    scope: str | None = ...,
    autouse: bool | None = ...,
    params: list[Any] | None = ...,
    ids: list[str] | Callable[[Any], str] | None = ...,
    name: str | None = ...,
) -> Callable[[Callable[P, R]], Callable[P, R]]: ...
def raises(
    exc: Type[R],
    match: str | Pattern[str] | None = ...,
    *,
    msg: str | None = ...,
) -> ContextManager[R]: ...

class mark:
    @staticmethod
    def parametrize(
        argnames: str | list[str],
        argvalues: list[Any] | tuple[Any, ...] | list[tuple[Any, ...]],
        *,
        ids: list[str] | Callable[[Any], str] | None = ...,
        indirect: bool | list[str] = ...,
        **kwargs: Any,
    ) -> Callable[[Callable[..., Any]], Callable[..., Any]]: ...
