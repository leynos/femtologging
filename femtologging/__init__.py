"""femtologging package."""

from __future__ import annotations

# Import the Rust extension packaged under this module's namespace first
# to keep imports at the top for linters.
from . import _femtologging_rs as rust  # type: ignore[attr-defined]
from .overflow_policy import OverflowPolicy
import ast
import logging
import sys
from dataclasses import dataclass
from typing import (
    Iterable,
    Mapping,
    MutableMapping,
    Sequence,
    TextIO,
    cast,
    overload,
)

hello = rust.hello  # type: ignore[attr-defined]
FemtoLogger = rust.FemtoLogger  # type: ignore[attr-defined]
get_logger = rust.get_logger  # type: ignore[attr-defined]
reset_manager = rust.reset_manager_py  # type: ignore[attr-defined]
FemtoHandler = rust.FemtoHandler  # type: ignore[attr-defined]
FemtoStreamHandler = rust.FemtoStreamHandler  # type: ignore[attr-defined]
FemtoFileHandler = rust.FemtoFileHandler  # type: ignore[attr-defined]
StreamHandlerBuilder = rust.StreamHandlerBuilder  # type: ignore[attr-defined]
FileHandlerBuilder = rust.FileHandlerBuilder  # type: ignore[attr-defined]
ConfigBuilder = rust.ConfigBuilder  # type: ignore[attr-defined]
LoggerConfigBuilder = rust.LoggerConfigBuilder  # type: ignore[attr-defined]
FormatterBuilder = rust.FormatterBuilder  # type: ignore[attr-defined]
HandlerConfigError = rust.HandlerConfigError  # type: ignore[attr-defined]
HandlerIOError = rust.HandlerIOError  # type: ignore[attr-defined]


# Mapping of string class names to handler builder classes. This mirrors the
# ``logging.config.dictConfig`` expectation of string references while steering
# users toward ``femtologging``'s builders.
_HANDLER_CLASS_MAP: dict[str, type[StreamHandlerBuilder] | type[FileHandlerBuilder]] = {
    "logging.StreamHandler": StreamHandlerBuilder,
    "femtologging.StreamHandler": StreamHandlerBuilder,
    "logging.FileHandler": FileHandlerBuilder,
    "femtologging.FileHandler": FileHandlerBuilder,
}


@dataclass
class BasicConfig:
    """Configuration parameters for basicConfig()."""

    level: str | int | None = None
    filename: str | None = None
    stream: TextIO | None = None
    force: bool = False
    handlers: Iterable[FemtoHandler] | None = None


@overload
def basicConfig(config: BasicConfig, /) -> None: ...


@overload
def basicConfig(**kwargs: object) -> None: ...


def basicConfig(config: BasicConfig | None = None, /, **kwargs: object) -> None:
    """Configure the root logger using the builder API.

    Parameters mirror ``logging.basicConfig`` but currently only a subset is
    supported. ``config`` may be a :class:`BasicConfig` instance; if provided,
    its values take precedence over individual parameters. ``level`` may be a
    string or numeric value understood by the standard :mod:`logging` module.
    ``filename`` configures a :class:`FemtoFileHandler`; otherwise a
    :class:`FemtoStreamHandler` targeting ``stderr`` is installed. ``stream``
    may be ``sys.stdout`` to redirect output. ``force`` removes any existing
    handlers from the root logger before applying the new configuration.
    ``handlers`` allows attaching pre‑constructed handlers directly.

    Parameters
    ----------
    config : BasicConfig, optional
        Configuration dataclass providing parameters for ``basicConfig``.

    Other Parameters
    ----------------
    level : str or int, optional
        Logging level.
    filename : str, optional
        File to write logs to.
    stream : TextIO, optional
        ``sys.stdout`` or ``sys.stderr``.
    force : bool, default False
        Remove any existing handlers before configuring.
    handlers : Iterable[FemtoHandler], optional
        Pre‑constructed handlers to attach.

    Examples
    --------
    Using a dataclass::

        cfg = BasicConfig(level="INFO")
        basicConfig(cfg)

    Using individual parameters::

        basicConfig(level="INFO")

    Notes
    -----
    ``format`` and ``datefmt`` are intentionally unsupported until formatter
    customisation is implemented.
    """
    allowed = {"level", "filename", "stream", "force", "handlers"}
    unknown = set(kwargs) - allowed
    if unknown:
        name = next(iter(unknown))
        raise TypeError(f"basicConfig() got an unexpected keyword argument {name!r}")

    level: str | int | None
    filename: str | None
    stream: TextIO | None
    force: bool
    handlers: Iterable[FemtoHandler] | None
    if config is not None:
        level = (
            config.level
            if config.level is not None
            else cast(str | int | None, kwargs.get("level"))
        )
        filename = (
            config.filename
            if config.filename is not None
            else cast(str | None, kwargs.get("filename"))
        )
        stream = (
            config.stream
            if config.stream is not None
            else cast(TextIO | None, kwargs.get("stream"))
        )
        force = bool(config.force)
        handlers = (
            config.handlers
            if config.handlers is not None
            else cast(Iterable[FemtoHandler] | None, kwargs.get("handlers"))
        )
    else:
        level = cast(str | int | None, kwargs.get("level"))
        filename = cast(str | None, kwargs.get("filename"))
        stream = cast(TextIO | None, kwargs.get("stream"))
        force = bool(kwargs.get("force", False))
        handlers = cast(Iterable[FemtoHandler] | None, kwargs.get("handlers"))

    _validate_basic_config_params(filename, stream, handlers)

    root = get_logger("root")
    if force:
        root.clear_handlers()

    _configure_handlers(root, handlers, filename, stream)

    _set_logger_level(root, level)


def _has_conflicting_handler_params(
    handlers: Iterable[FemtoHandler] | None,
    filename: str | None,
    stream: TextIO | None,
) -> bool:
    """Check if handlers conflict with filename or stream parameters."""
    return handlers is not None and (filename is not None or stream is not None)


def _validate_basic_config_params(
    filename: str | None,
    stream: TextIO | None,
    handlers: Iterable[FemtoHandler] | None,
) -> None:
    """Validate ``basicConfig`` parameters."""
    if filename and stream:
        raise ValueError("Cannot specify both `filename` and `stream`")

    if _has_conflicting_handler_params(handlers, filename, stream):
        msg = "Cannot specify `handlers` with `filename` or `stream`"
        raise ValueError(msg)

    if stream not in (None, sys.stdout, sys.stderr):
        raise ValueError(
            f"stream must be sys.stdout or sys.stderr, got {type(stream)!r}: {stream!r}"
        )


def _configure_handlers(
    root: FemtoLogger,
    handlers: Iterable[FemtoHandler] | None,
    filename: str | None,
    stream: TextIO | None,
) -> None:
    """Attach or build handlers for the root logger."""
    if handlers is not None:
        for h in handlers:
            root.add_handler(h)
    else:
        _build_and_configure_handler(filename, stream)


def _build_and_configure_handler(
    filename: str | None,
    stream: TextIO | None,
) -> None:
    """Build a handler via the builder API and install it."""
    builder = ConfigBuilder()

    handler_id = "basic_config_handler"
    handler = _create_handler_builder(filename, stream)
    builder.with_handler(handler_id, handler)

    logger_cfg = LoggerConfigBuilder().with_handlers([handler_id])
    builder.with_root_logger(logger_cfg)

    builder.build_and_init()


def _create_handler_builder(
    filename: str | None,
    stream: TextIO | None,
) -> FileHandlerBuilder | StreamHandlerBuilder:
    """Create a handler builder for ``basicConfig``."""
    if filename:
        return FileHandlerBuilder(filename)
    if stream is sys.stdout:
        return StreamHandlerBuilder.stdout()
    return StreamHandlerBuilder.stderr()


def _set_logger_level(root: FemtoLogger, level: str | int | None) -> None:
    """Set the root logger level if provided."""
    if level is not None:
        lvl = logging.getLevelName(level) if isinstance(level, int) else level
        root.set_level(lvl)


def _coerce_args(args: object) -> list[object]:
    """Convert ``args`` into a list for handler construction.

    ``dictConfig`` accepts either a sequence or a string representing a Python
    literal. The string form is evaluated safely using ``ast.literal_eval`` so
    configuration loaded from text formats (e.g. JSON) can express tuples.
    """
    if isinstance(args, str):
        try:
            args = ast.literal_eval(args)
        except (ValueError, SyntaxError) as exc:
            raise ValueError(f"invalid args: {args}") from exc
    if args is None:
        return []
    if not isinstance(args, Sequence):
        raise ValueError("args must be a sequence")
    return list(args)


def _coerce_kwargs(kwargs: object) -> dict[str, object]:
    """Convert ``kwargs`` into a dictionary for handler construction."""
    if isinstance(kwargs, str):
        try:
            kwargs = ast.literal_eval(kwargs)
        except (ValueError, SyntaxError) as exc:
            raise ValueError(f"invalid kwargs: {kwargs}") from exc
    if kwargs is None:
        return {}
    if not isinstance(kwargs, Mapping):
        raise ValueError("kwargs must be a mapping")
    return dict(kwargs)


def _resolve_handler_class(
    name: str,
) -> type[StreamHandlerBuilder] | type[FileHandlerBuilder]:
    """Return the builder class for ``name`` or raise ``ValueError``."""
    cls = _HANDLER_CLASS_MAP.get(name)
    if cls is None:
        raise ValueError(f"unsupported handler class {name!r}")
    return cls


def _build_handler_from_dict(
    hid: str, data: MutableMapping[str, object]
) -> FileHandlerBuilder | StreamHandlerBuilder:
    """Create a handler builder from ``dictConfig`` handler data."""
    cls_name = data.get("class")
    if not isinstance(cls_name, str):
        raise ValueError(f"handler {hid!r} missing class")
    builder_cls = _resolve_handler_class(cls_name)
    args = _coerce_args(data.get("args"))
    kwargs = _coerce_kwargs(data.get("kwargs"))
    try:
        builder = builder_cls(*args, **kwargs)
    except Exception as exc:  # pragma: no cover - constructor errors propagated
        raise ValueError(f"failed to construct handler {hid!r}: {exc}") from exc
    if data.get("level") is not None:
        raise ValueError("handler level is not supported")
    if data.get("filters"):
        raise ValueError("handler filters are not supported")
    if fmt := data.get("formatter"):
        builder = builder.with_formatter(fmt)
    return builder


def _build_logger_from_dict(
    name: str, data: MutableMapping[str, object]
) -> LoggerConfigBuilder:
    """Create a ``LoggerConfigBuilder`` from ``dictConfig`` logger data."""
    if data.get("filters"):
        raise ValueError("filters are not supported")
    builder = LoggerConfigBuilder()
    if lvl := data.get("level"):
        builder = builder.with_level(lvl)
    if handlers := data.get("handlers"):
        builder = builder.with_handlers(handlers)
    if prop := data.get("propagate"):
        builder = builder.with_propagate(bool(prop))
    return builder


def dictConfig(config: Mapping[str, object]) -> None:
    """Configure logging using a ``dictConfig``-style dictionary."""
    if config.get("incremental"):
        raise ValueError("incremental configuration is not supported")
    version = int(config.get("version", 1))
    if version != 1:
        raise ValueError(f"unsupported configuration version {version}")
    builder = ConfigBuilder().with_version(version)
    if config.get("disable_existing_loggers"):
        builder.with_disable_existing_loggers(True)

    formatters = config.get("formatters", {})
    if not isinstance(formatters, Mapping):
        raise ValueError("formatters must be a mapping")
    for fid, fcfg in formatters.items():
        if not isinstance(fcfg, Mapping):
            raise ValueError("formatter config must be a mapping")
        fb = FormatterBuilder()
        if fmt := fcfg.get("format"):
            fb = fb.with_format(fmt)
        if datefmt := fcfg.get("datefmt"):
            fb = fb.with_datefmt(datefmt)
        builder.with_formatter(fid, fb)

    if config.get("filters"):
        raise ValueError("filters are not supported")

    handlers = config.get("handlers", {})
    if not isinstance(handlers, Mapping):
        raise ValueError("handlers must be a mapping")
    for hid, hcfg in handlers.items():
        if not isinstance(hcfg, MutableMapping):
            raise ValueError("handler config must be a mapping")
        builder.with_handler(hid, _build_handler_from_dict(hid, hcfg))

    loggers = config.get("loggers", {})
    if not isinstance(loggers, Mapping):
        raise ValueError("loggers must be a mapping")
    for name, lcfg in loggers.items():
        if not isinstance(lcfg, MutableMapping):
            raise ValueError("logger config must be a mapping")
        builder.with_logger(name, _build_logger_from_dict(name, lcfg))

    root = config.get("root")
    if not isinstance(root, MutableMapping):
        raise ValueError("root logger configuration is required")
    builder.with_root_logger(_build_logger_from_dict("root", root))

    builder.build_and_init()


__all__ = [
    "FemtoHandler",
    "FemtoLogger",
    "get_logger",
    "reset_manager",
    "FemtoStreamHandler",
    "FemtoFileHandler",
    "StreamHandlerBuilder",
    "FileHandlerBuilder",
    "ConfigBuilder",
    "LoggerConfigBuilder",
    "FormatterBuilder",
    "HandlerConfigError",
    "HandlerIOError",
    "OverflowPolicy",
    "BasicConfig",
    "basicConfig",
    "dictConfig",
    "hello",
]
