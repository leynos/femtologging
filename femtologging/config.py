"""Configuration via loggingâ€‘style dictionaries.

This module implements :func:`dictConfig`, a restricted variant of
``logging.config.dictConfig``. Only a subset of the standard schema is
recognized: ``filters`` sections, handler ``level`` attributes, and
incremental configuration are unsupported.

String level parameters accept case-insensitive names: "TRACE", "DEBUG",
"INFO", "WARN", "WARNING", "ERROR", and "CRITICAL". "WARN" and "WARNING"
are equivalent.

Example
-------
>>> dictConfig({
...     "version": 1,
...     "handlers": {"h": {"class": "femtologging.StreamHandler"}},
...     "root": {"level": "INFO", "handlers": ["h"]},
... })

The ``dictConfig`` format does not support ``filters`` and will raise ``ValueError`` if a ``filters`` section is provided. To attach filters, use the builder API:

    cb = (
        ConfigBuilder()
        .with_filter("lvl", LevelFilterBuilder().with_max_level("INFO"))
        .with_logger("core", LoggerConfigBuilder().with_filters(["lvl"]))
        .with_root_logger(LoggerConfigBuilder().with_level("INFO"))
    )
    cb.build_and_init()
"""

from __future__ import annotations

import ast
from dataclasses import dataclass
from typing import Any, Callable, Final, Mapping, Sequence, cast
import typing as _typing

if _typing.TYPE_CHECKING:
    from ._femtologging_rs import (  # noqa: F401
        LevelFilterBuilder as LevelFilterBuilder,
        NameFilterBuilder as NameFilterBuilder,
    )

from . import _femtologging_rs as rust
from .overflow_policy import OverflowPolicy

rust = cast(Any, rust)
HandlerConfigError: type[Exception] = getattr(rust, "HandlerConfigError", Exception)
HandlerIOError: type[Exception] = getattr(rust, "HandlerIOError", Exception)

StreamHandlerBuilder = rust.StreamHandlerBuilder
FileHandlerBuilder = rust.FileHandlerBuilder
ConfigBuilder = rust.ConfigBuilder
LoggerConfigBuilder = rust.LoggerConfigBuilder
FormatterBuilder = rust.FormatterBuilder
LevelFilterBuilder = rust.LevelFilterBuilder
NameFilterBuilder = rust.NameFilterBuilder


_HANDLER_CLASS_MAP: Final[dict[str, object]] = {
    "logging.StreamHandler": StreamHandlerBuilder,
    "femtologging.StreamHandler": StreamHandlerBuilder,
    "logging.FileHandler": FileHandlerBuilder,
    "femtologging.FileHandler": FileHandlerBuilder,
}


def _evaluate_string_safely(value: str, context: str) -> object:
    """Safely evaluate a string ``value`` using ``ast.literal_eval``."""
    try:
        return ast.literal_eval(value)
    except (ValueError, SyntaxError) as exc:
        raise ValueError(f"invalid {context}: {value}") from exc


def _validate_mapping_type(value: object, name: str) -> Mapping[object, object]:
    """Ensure ``value`` is a mapping and not bytes-like."""
    if isinstance(value, (bytes, bytearray)) or not isinstance(value, Mapping):
        raise ValueError(f"{name} must be a mapping")
    return cast(Mapping[object, object], value)


def _validate_no_bytes(value: object, name: str) -> None:
    """Reject ``bytes`` or ``bytearray`` for ``value``."""
    if isinstance(value, (bytes, bytearray)):
        raise ValueError(f"{name} must not be bytes or bytearray")


def _validate_string_keys(
    mapping: Mapping[object, object], name: str
) -> Mapping[str, object]:
    """Ensure all keys in ``mapping`` are strings."""
    for key in mapping:
        if not isinstance(key, str):
            raise ValueError(f"{name} keys must be strings")
    return cast(Mapping[str, object], mapping)


def _coerce_args(args: object, ctx: str) -> list[object]:
    """Convert ``args`` into a list for handler construction."""
    if isinstance(args, str):
        args = _evaluate_string_safely(args, f"{ctx} args")
    if args is None:
        return []
    _validate_no_bytes(args, f"{ctx} args")
    if not isinstance(args, Sequence):
        raise ValueError(f"{ctx} args must be a sequence")
    return list(cast(Sequence[object], args))


def _coerce_kwargs(kwargs: object, ctx: str) -> dict[str, object]:
    """Convert ``kwargs`` into a dictionary for handler construction."""
    if isinstance(kwargs, str):
        kwargs = _evaluate_string_safely(kwargs, f"{ctx} kwargs")
    if kwargs is None:
        return {}
    mapping = _validate_mapping_type(kwargs, f"{ctx} kwargs")
    mapping = _validate_string_keys(mapping, f"{ctx} kwargs")
    result: dict[str, object] = {}
    for key, value in mapping.items():
        _validate_no_bytes(value, f"{ctx} kwargs values")
        result[key] = value
    return result


def _resolve_handler_class(name: str) -> object:
    """Return the builder class for ``name`` or raise ``ValueError``."""
    cls = _HANDLER_CLASS_MAP.get(name)
    if cls is None:
        raise ValueError(f"unsupported handler class {name!r}")
    return cls


def _validate_handler_keys(hid: str, data: Mapping[str, object]) -> None:
    """Validate that ``data`` contains only supported handler keys."""
    allowed = {"class", "level", "filters", "args", "kwargs", "formatter"}
    unknown = set(data.keys()) - allowed
    if unknown:
        raise ValueError(f"handler {hid!r} has unsupported keys: {sorted(unknown)!r}")


def _validate_handler_class(hid: str, cls_name: object) -> str:
    """Ensure a string handler class name is provided."""
    if not isinstance(cls_name, str):
        raise ValueError(f"handler {hid!r} missing class")
    return cls_name


def _validate_unsupported_features(data: Mapping[str, object]) -> None:
    """Reject handler features not yet implemented."""
    if "level" in data:
        raise ValueError("handler level is not supported")
    if "filters" in data:
        raise ValueError("handler filters are not supported")


def _validate_handler_config(
    hid: str, data: Mapping[str, object]
) -> tuple[str, list[object], dict[str, object], object | None]:
    """Validate handler ``data`` and return construction parameters."""
    _validate_handler_keys(hid, data)
    cls_name = _validate_handler_class(hid, data.get("class"))
    _validate_unsupported_features(data)
    ctx = f"handler {hid!r}"
    args = _coerce_args(data.get("args"), ctx)
    kwargs = _coerce_kwargs(data.get("kwargs"), ctx)
    return cls_name, args, kwargs, data.get("formatter")


def _create_handler_instance(
    hid: str, cls_name: str, args: list[object], kwargs: dict[str, object]
) -> object:
    """Instantiate a handler builder and wrap constructor errors."""
    builder_cls = _resolve_handler_class(cls_name)
    try:
        args_t = tuple(args)
        kwargs_d = dict(kwargs)
        return cast(Any, builder_cls)(*args_t, **kwargs_d)  # pyright: ignore[reportCallIssue]
    except (TypeError, ValueError, HandlerConfigError, HandlerIOError) as exc:
        raise ValueError(f"failed to construct handler {hid!r}: {exc}") from exc


def _build_handler_from_dict(hid: str, data: Mapping[str, object]) -> object:
    """Create a handler builder from ``dictConfig`` handler data."""
    cls_name, args, kwargs, fmt = _validate_handler_config(hid, data)
    builder = cast(Any, _create_handler_instance(hid, cls_name, args, kwargs))
    if fmt is not None:
        if not isinstance(fmt, str):
            raise ValueError("formatter must be a string")
        builder = builder.with_formatter(fmt)
    return builder


def _validate_logger_handlers(handlers_obj: object) -> list[str]:
    """Validate logger ``handlers`` list and return it."""
    if not isinstance(handlers_obj, (list, tuple)):
        raise ValueError("logger handlers must be a list or tuple of strings")
    handlers_seq = cast(Sequence[object], handlers_obj)
    if not all(isinstance(h, str) for h in handlers_seq):
        raise ValueError("logger handlers must be a list or tuple of strings")
    return list(cast(Sequence[str], handlers_seq))


def _validate_logger_config_keys(name: str, data: Mapping[str, object]) -> None:
    """Ensure ``data`` uses only supported logger keys."""
    allowed = {"level", "handlers", "propagate", "filters"}
    unknown = set(data.keys()) - allowed
    if unknown:
        raise ValueError(f"logger {name!r} has unsupported keys: {sorted(unknown)!r}")
    if "filters" in data:
        raise ValueError("filters are not supported")


def _validate_propagate_value(value: object) -> bool:
    """Validate the ``propagate`` value for a logger."""
    if not isinstance(value, bool):
        raise ValueError("logger propagate must be a bool")
    return value


def _build_logger_from_dict(name: str, data: Mapping[str, object]) -> object:
    """Create a ``LoggerConfigBuilder`` from ``dictConfig`` logger data."""
    _validate_logger_config_keys(name, data)
    builder = LoggerConfigBuilder()
    if "level" in data:
        builder = builder.with_level(data["level"])
    if "handlers" in data:
        handlers = _validate_logger_handlers(data["handlers"])
        builder = builder.with_handlers(handlers)
    if "propagate" in data:
        propagate = _validate_propagate_value(data["propagate"])
        builder = builder.with_propagate(propagate)
    return builder


def _validate_dict_config(config: Mapping[str, object]) -> int:
    """Validate top-level configuration and return the version."""
    if "incremental" in config:
        raise ValueError("incremental configuration is not supported")
    version = int(cast(int, config.get("version", 1)))
    if version != 1:
        raise ValueError(f"unsupported configuration version {version}")
    if "filters" in config:
        raise ValueError("filters are not supported")
    return version


def _create_config_builder(version: int, config: Mapping[str, object]) -> object:
    """Initialize a ``ConfigBuilder`` with global options."""
    cb = ConfigBuilder()
    builder = cb.with_version(version)
    if "disable_existing_loggers" in config:
        value = config["disable_existing_loggers"]
        if not isinstance(value, bool):
            raise ValueError("disable_existing_loggers must be a bool")
        builder = builder.with_disable_existing_loggers(value)
    return builder


def _validate_formatter_field(
    fcfg: Mapping[str, object], field: str, field_type: str
) -> str | None:
    """Return the string value for ``field`` or ``None`` if absent."""
    if field not in fcfg:
        return None
    value = fcfg[field]
    if not isinstance(value, str):
        raise ValueError(f"formatter '{field_type}' must be a string")
    return value


def _build_formatter(fcfg: Mapping[str, object]) -> object:
    """Build a :class:`FormatterBuilder` from configuration."""
    allowed = {"format", "datefmt"}
    unknown = set(fcfg.keys()) - allowed
    if unknown:
        raise ValueError(f"formatter has unsupported keys: {sorted(unknown)!r}")
    fb = FormatterBuilder()
    fmt = _validate_formatter_field(fcfg, "format", "format")
    if fmt is not None:
        fb = fb.with_format(fmt)
    datefmt = _validate_formatter_field(fcfg, "datefmt", "datefmt")
    if datefmt is not None:
        fb = fb.with_datefmt(datefmt)
    return fb


def _validate_section_mapping(section: object, name: str) -> Mapping[str, object]:
    """Ensure a configuration ``section`` is a mapping."""
    return cast(Mapping[str, object], _validate_mapping_type(section, name))


@dataclass(frozen=True)
class SectionProcessor:
    """Configuration for :func:`_process_config_section`."""

    section: str
    builder_method: str
    build_func: Callable[[str, Mapping[str, object]], object]
    err_tmpl: str | None = None


def _process_config_section(
    builder: Any, config: Mapping[str, object], processor: SectionProcessor
) -> None:
    """Generic processor for formatter, handler, and logger sections."""
    mapping = cast(
        Mapping[object, object],
        _validate_section_mapping(config.get(processor.section, {}), processor.section),
    )
    method = getattr(builder, processor.builder_method)
    for key, cfg in mapping.items():
        if not isinstance(key, str):
            if processor.err_tmpl is None:
                raise ValueError(f"{processor.section[:-1]} ids must be strings")
            raise ValueError(processor.err_tmpl.format(name=repr(key)))
        method(
            key,
            processor.build_func(
                key,
                _validate_section_mapping(cfg, f"{processor.section[:-1]} config"),
            ),
        )


def _process_formatters(builder: Any, config: Mapping[str, object]) -> None:
    """Attach formatter builders to ``builder``."""
    _process_config_section(
        builder,
        config,
        SectionProcessor(
            "formatters", "with_formatter", lambda fid, m: _build_formatter(m)
        ),
    )


def _process_handlers(builder: Any, config: Mapping[str, object]) -> None:
    """Attach handler builders to ``builder``."""
    _process_config_section(
        builder,
        config,
        SectionProcessor(
            "handlers", "with_handler", lambda hid, m: _build_handler_from_dict(hid, m)
        ),
    )


def _process_loggers(builder: Any, config: Mapping[str, object]) -> None:
    """Attach logger configurations to ``builder``."""
    _process_config_section(
        builder,
        config,
        SectionProcessor(
            "loggers",
            "with_logger",
            lambda name, m: _build_logger_from_dict(name, m),
            err_tmpl="loggers section key {name} must be a string",
        ),
    )


def _process_root_logger(builder: Any, config: Mapping[str, object]) -> None:
    """Configure the root logger."""
    if "root" not in config:
        raise ValueError("root logger configuration is required")
    root = config["root"]
    if not isinstance(root, Mapping):
        raise ValueError("root logger configuration must be a mapping")
    builder.with_root_logger(
        _build_logger_from_dict("root", cast(Mapping[str, object], root))
    )


def dictConfig(config: Mapping[str, object]) -> None:
    """Configure logging using a ``dictConfig``â€‘style dictionary.

    Parameters
    ----------
    config : Mapping[str, object]
        A dictionary compatible with :mod:`logging.config`. Supported keys are
        ``version``, ``disable_existing_loggers``, ``formatters``, ``handlers``,
        ``loggers``, and ``root``. Unsupported features (e.g., ``filters``,
        handler ``level``) raise ``ValueError``.

    Raises
    ------
    ValueError
        If the configuration uses unsupported features or invalid schemas.

    Examples
    --------
    >>> dictConfig({
    ...     "version": 1,
    ...     "handlers": {"h": {"class": "femtologging.StreamHandler"}},
    ...     "root": {"level": "INFO", "handlers": ["h"]},
    ... })
    """

    version = _validate_dict_config(config)
    builder = cast(Any, _create_config_builder(version, config))
    _process_formatters(builder, config)
    _process_handlers(builder, config)
    _process_loggers(builder, config)
    _process_root_logger(builder, config)
    builder.build_and_init()


__all__ = [
    "ConfigBuilder",
    "LoggerConfigBuilder",
    "FormatterBuilder",
    "StreamHandlerBuilder",
    "FileHandlerBuilder",
    "LevelFilterBuilder",
    "NameFilterBuilder",
    "dictConfig",
    "OverflowPolicy",
]
