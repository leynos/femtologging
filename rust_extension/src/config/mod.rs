//! Configuration builders for femtologging.

mod build;
mod py;
mod types;

#[allow(unused_imports)] // re-export for external consumers
pub use types::{ConfigBuilder, ConfigError, FormatterBuilder, LoggerConfigBuilder};

#[cfg(test)]
mod config_tests;
