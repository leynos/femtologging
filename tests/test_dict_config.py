"""Tests for femtologging.dictConfig integration and behaviour."""

from __future__ import annotations

from pathlib import Path
import time

import pytest
from pytest_bdd import given, parsers, scenarios, then, when

from femtologging import ConfigBuilder, dictConfig, get_logger, reset_manager
from femtologging.config import _process_formatters, _process_handlers

scenarios("features/dict_config.feature")


@given("the logging system is reset")
def reset_logging() -> None:
    reset_manager()


@when("I configure dictConfig with a stream handler")
def configure_dict_config() -> None:
    cfg = {
        "version": 1,
        "handlers": {"h": {"class": "femtologging.StreamHandler"}},
        "root": {"level": "INFO", "handlers": ["h"]},
    }
    dictConfig(cfg)


@then(parsers.parse('logging "{msg}" at "{level}" from root matches snapshot'))
def log_matches_snapshot(msg: str, level: str, snapshot) -> None:
    logger = get_logger("root")
    assert logger.log(level, msg) == snapshot


@then("calling dictConfig with incremental true raises ValueError")
def dict_config_incremental_fails() -> None:
    with pytest.raises(ValueError, match="incremental configuration is not supported"):
        dictConfig({"version": 1, "incremental": True, "root": {}})


@when(
    parsers.parse('I configure dictConfig with handler class "{cls}"'),
    target_fixture="config_error",
)
def configure_with_handler_class(cls: str) -> BaseException:
    cfg = {
        "version": 1,
        "handlers": {"h": {"class": cls}},
        "root": {"level": "INFO", "handlers": ["h"]},
    }
    with pytest.raises(ValueError) as exc:
        dictConfig(cfg)
    return exc.value


@then("dictConfig raises ValueError")
def dict_config_raises_value_error(config_error: BaseException) -> None:
    assert isinstance(config_error, ValueError)


def test_dict_config_file_handler_args_kwargs(tmp_path: Path) -> None:
    """Verify args and kwargs are evaluated for handler construction."""
    reset_manager()
    path = tmp_path / "out.log"
    cfg = {
        "version": 1,
        "handlers": {
            "f": {
                "class": "femtologging.FileHandler",
                "args": f"('{path}',)",
                "kwargs": "{}",
            }
        },
        "root": {"level": "INFO", "handlers": ["f"]},
    }
    dictConfig(cfg)
    logger = get_logger("root")
    logger.log("INFO", "file")
    time.sleep(0.05)
    assert path.exists()
    contents = path.read_text()
    assert "file" in contents


def test_dict_config_args_reject_bytes() -> None:
    reset_manager()
    cfg = {
        "version": 1,
        "handlers": {"h": {"class": "femtologging.StreamHandler", "args": b"bytes"}},
        "root": {"level": "INFO", "handlers": ["h"]},
    }
    with pytest.raises(
        ValueError, match="handler 'h' args must not be bytes or bytearray"
    ):
        dictConfig(cfg)


def test_dict_config_kwargs_reject_bytes_value() -> None:
    reset_manager()
    cfg = {
        "version": 1,
        "handlers": {
            "h": {
                "class": "femtologging.StreamHandler",
                "kwargs": {"path": b"oops"},
            }
        },
        "root": {"level": "INFO", "handlers": ["h"]},
    }
    with pytest.raises(
        ValueError,
        match="handler 'h' kwargs values must not be bytes or bytearray",
    ):
        dictConfig(cfg)


def test_process_formatters_apply_to_handlers() -> None:
    builder = ConfigBuilder().with_version(1)
    cfg = {
        "formatters": {"f": {"format": "%(message)s", "datefmt": "%H:%M"}},
        "handlers": {
            "h": {
                "class": "femtologging.StreamHandler",
                "formatter": "f",
            }
        },
    }
    _process_formatters(builder, cfg)
    _process_handlers(builder, cfg)
    state = builder.as_dict()
    fmt = state["formatters"]["f"]
    assert fmt["format"] == "%(message)s"
    assert fmt["datefmt"] == "%H:%M"
    assert state["handlers"]["h"]["formatter_id"] == "f"


def test_dict_config_handler_filters_presence() -> None:
    reset_manager()
    cfg = {
        "version": 1,
        "handlers": {
            "h": {
                "class": "femtologging.StreamHandler",
                "filters": [],
            }
        },
        "root": {"level": "INFO", "handlers": ["h"]},
    }
    with pytest.raises(ValueError, match="handler filters are not supported"):
        dictConfig(cfg)


def test_dict_config_logger_filters_presence() -> None:
    reset_manager()
    cfg = {
        "version": 1,
        "loggers": {"a": {"filters": []}},
        "root": {"handlers": []},
    }
    with pytest.raises(ValueError, match="filters are not supported"):
        dictConfig(cfg)


@pytest.mark.parametrize(
    "config",
    [
        {"version": 2, "root": {}},
        {
            "version": 1,
            "handlers": {"h": {"class": "unknown"}},
            "root": {"handlers": ["h"]},
        },
        {"version": 1, "filters": {"f": {}}, "root": {}},
    ],
    ids=["version-unsupported", "handler-class-unknown", "filters-unsupported"],
)
def test_dict_config_invalid_configs(config: dict[str, object]) -> None:
    """Invalid configurations raise ``ValueError``."""
    reset_manager()
    with pytest.raises(ValueError):
        dictConfig(config)
