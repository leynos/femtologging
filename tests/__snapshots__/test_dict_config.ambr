# serializer version: 1
# name: test_configure_root_logger_via_dictconfig
  'root [INFO] hello'
# ---
